### PR TITLE
[CARBONDATA-2872] Added Spark FileFormat interface implementation in Carbon

### PR DESCRIPTION
**Pull request by ravipesala**
Added new package carbondata-spark-datasource under <carbondata>/integration/spark-datasource
It contains the implementation of Spark's FileFormat and user can use `carbon` as format in spark
For example
```
create table test_table(c1 string, c2 int) using carbon
or
dataframe.write.format("carbon").saveAsTable("test_table")
```
There are few classes moved to this datasource package as part of refactoring and spark2 and spark-common packages now depends on spark-datasource package.

Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [ ] Any interfaces changed?
 
 - [ ] Any backward compatibility impacted?
 
 - [ ] Document update required?

 - [ ] Testing done
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
        - How it is tested? Please attach test report.
        - Is it a performance related change? Please attach the performance test report.
        - Any additional information to help reviewers in testing this change.
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

### CODE REVIEW

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/6316/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7966/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6690/

File: core/src/main/java/org/apache/carbondata/core/readcommitter/LatestFilesReadCommittedScope.java [hunk 1]
```diff
@@ -66,7 +68,23 @@ public LatestFilesReadCommittedScope(String path, String segmentId) {
...
   public LatestFilesReadCommittedScope(String path) {
-    this(path, null);
+    this(path, (String) null);
+  }
+
+  /**
+   * a new constructor with path
+   *
+   * @param path carbon file path
+   */
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/readcommitter/LatestFilesReadCommittedScope.java]: missing parameter doc

File: core/src/main/java/org/apache/carbondata/core/readcommitter/LatestFilesReadCommittedScope.java [hunk 2]
```diff
@@ -36,12 +36,14 @@
  */
 @InterfaceAudience.Internal
 @InterfaceStability.Stable
-public class LatestFilesReadCommittedScope implements ReadCommittedScope {
+public class LatestFilesReadCommittedScope
+    implements ReadCommittedScope {
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/readcommitter/LatestFilesReadCommittedScope.java]: no need to modify

File: core/src/main/java/org/apache/carbondata/core/readcommitter/LatestFilesReadCommittedScope.java [hunk 3]
```diff
@@ -66,7 +68,23 @@ public LatestFilesReadCommittedScope(String path, String segmentId) {
...
+   * @param path carbon file path
+   */
+  public LatestFilesReadCommittedScope(String path, String[] subFolders) {
+    Objects.requireNonNull(path);
+    this.carbonFilePath = path;
+    this.subFolders = subFolders;
+    try {
+      takeCarbonIndexFileSnapShot();
+    } catch (IOException ex) {
+      throw new RuntimeException("Error while taking index snapshot", ex);
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/readcommitter/LatestFilesReadCommittedScope.java]: Can we use more specific exception

File: core/src/main/java/org/apache/carbondata/core/readcommitter/LatestFilesReadCommittedScope.java [hunk 4]
```diff
@@ -66,7 +68,23 @@ public LatestFilesReadCommittedScope(String path, String segmentId) {
...
+
+  /**
+   * a new constructor with path
+   *
+   * @param path carbon file path
+   */
+  public LatestFilesReadCommittedScope(String path, String[] subFolders) {
+    Objects.requireNonNull(path);
+    this.carbonFilePath = path;
+    this.subFolders = subFolders;
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/readcommitter/LatestFilesReadCommittedScope.java]: suggest change `subFolders` to `dataFolders` and add javadoc to explain it contains carbondata files and carbonindex files in these folders

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/6319/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6693/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7969/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/6324/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6699/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7975/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/6326/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6701/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7977/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/6332/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7983/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/6336/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6709/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7985/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/6340/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7989/

File: core/src/main/java/org/apache/carbondata/core/readcommitter/LatestFilesReadCommittedScope.java (hunk 3 above)
Comment (ravipesala) [reply core/src/main/java/org/apache/carbondata/core/readcommitter/LatestFilesReadCommittedScope.java]: Not required handling here, so removed.

File: core/src/main/java/org/apache/carbondata/core/readcommitter/LatestFilesReadCommittedScope.java (hunk 2 above)
Comment (ravipesala) [reply core/src/main/java/org/apache/carbondata/core/readcommitter/LatestFilesReadCommittedScope.java]: ok

File: core/src/main/java/org/apache/carbondata/core/readcommitter/LatestFilesReadCommittedScope.java (hunk 1 above)
Comment (ravipesala) [reply core/src/main/java/org/apache/carbondata/core/readcommitter/LatestFilesReadCommittedScope.java]: ok

File: core/src/main/java/org/apache/carbondata/core/readcommitter/LatestFilesReadCommittedScope.java (hunk 4 above)
Comment (ravipesala) [reply core/src/main/java/org/apache/carbondata/core/readcommitter/LatestFilesReadCommittedScope.java]: ok

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/6349/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7997/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/6351/

File: integration/spark-datasource/src/main/scala/org/apache/spark/sql/carbondata/execution/datasources/CarbonSparkDataSourceUtil.scala [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+            new OrExpression(lhsFilter, rhsFilter)
+          }
+        case sources.StringStartsWith(name, value) if value.length > 0 =>
+          Some(new StartsWithExpression(getCarbonExpression(name),
+            getCarbonLiteralExpression(name, value)))
+        case _ => None
+      }
+    }
+
+    def getCarbonExpression(name: String) = {
```
Comment (KanakaKumar) [inline integration/spark-datasource/src/main/scala/org/apache/spark/sql/carbondata/execution/datasources/CarbonSparkDataSourceUtil.scala]: Remove duplicate methods (CarbonFilters also has)

File: integration/spark-datasource/src/main/scala/org/apache/spark/sql/carbondata/execution/datasources/CarbonSparkDataSourceUtil.scala [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+      builder.withBlockletSize(blockletSize.get)
+    }
+    builder.enableLocalDictionary(options.getOrElse(CarbonCommonConstants.LOCAL_DICTIONARY_ENABLE,
+      CarbonCommonConstants.LOCAL_DICTIONARY_ENABLE_DEFAULT).toBoolean)
+    builder.localDictionaryThreshold(
+      options.getOrElse(CarbonCommonConstants.LOCAL_DICTIONARY_THRESHOLD,
+        CarbonCommonConstants.LOCAL_DICTIONARY_THRESHOLD_DEFAULT).toInt)
+    builder.sortBy(
+      options.get(CarbonCommonConstants.SORT_COLUMNS).map(_.split(",").map(_.trim)).orNull)
+    builder.isTransactionalTable(false)
```
Comment (KanakaKumar) [inline integration/spark-datasource/src/main/scala/org/apache/spark/sql/carbondata/execution/datasources/CarbonSparkDataSourceUtil.scala]: duplicate line

File: integration/spark-datasource/src/main/scala/org/apache/spark/sql/carbondata/execution/datasources/CarbonSparkDataSourceUtil.scala [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+    builder.uniqueIdentifier(System.currentTimeMillis())
```
Comment (KanakaKumar) [inline integration/spark-datasource/src/main/scala/org/apache/spark/sql/carbondata/execution/datasources/CarbonSparkDataSourceUtil.scala]: Support column_meta_cache, cache_level also

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6722/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7999/

File: integration/spark-datasource/src/main/scala/org/apache/spark/sql/carbondata/execution/datasources/readsupport/SparkUnsafeRowReadSuport.scala [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+import org.apache.spark.sql.types.StructType
+
+import org.apache.carbondata.core.metadata.schema.table.CarbonTable
+import org.apache.carbondata.core.metadata.schema.table.column.CarbonColumn
+import org.apache.carbondata.hadoop.readsupport.impl.DictionaryDecodeReadSupport
+
+/**
+ * Read support class which converts carbon row array format to sparks Internal row.
+ */
+class SparkUnsafeRowReadSuport(requiredSchema: StructType)
```
Comment (KanakaKumar) [inline integration/spark-datasource/src/main/scala/org/apache/spark/sql/carbondata/execution/datasources/readsupport/SparkUnsafeRowReadSuport.scala]: Directly implement CarbonReadSupport

File: integration/spark-datasource/pom.xml [hunk 9]
```diff
@@ -0,0 +1,279 @@
...
+      <id>spark-2.2</id>
+      <activation>
+        <activeByDefault>true</activeByDefault>
+      </activation>
+      <properties>
+        <spark.version>2.2.1</spark.version>
+        <scala.binary.version>2.11</scala.binary.version>
+        <scala.version>2.11.8</scala.version>
+      </properties>
+      <build>
```
[the remark this answers is not in the document]
Comment (ravipesala) [reply integration/spark-datasource/pom.xml]: ok

File: integration/spark-datasource/src/main/scala/org/apache/spark/sql/carbondata/execution/datasources/CarbonSparkDataSourceUtil.scala (hunk 5 above)
Comment (ravipesala) [reply integration/spark-datasource/src/main/scala/org/apache/spark/sql/carbondata/execution/datasources/CarbonSparkDataSourceUtil.scala]: ok

File: integration/spark-datasource/src/main/scala/org/apache/spark/sql/carbondata/execution/datasources/CarbonSparkDataSourceUtil.scala (hunk 7 above)
Comment (ravipesala) [reply integration/spark-datasource/src/main/scala/org/apache/spark/sql/carbondata/execution/datasources/CarbonSparkDataSourceUtil.scala]: ok

File: integration/spark-datasource/src/main/scala/org/apache/spark/sql/carbondata/execution/datasources/SparkCarbonFileFormat.scala [hunk 10]
```diff
@@ -0,0 +1,400 @@
...
+      case None => None
+    }
+    val format: CarbonFileInputFormat[Object] = new CarbonFileInputFormat[Object]
+    val broadcastedHadoopConf =
+      sparkSession.sparkContext.broadcast(new SerializableConfiguration(hadoopConf))
+
+    file: PartitionedFile => {
+      assert(file.partitionValues.numFields == partitionSchema.size)
+
+      if (!(file.filePath.endsWith(CarbonTablePath.INDEX_FILE_EXT) ||
```
[the remark this answers is not in the document]
Comment (ravipesala) [reply integration/spark-datasource/src/main/scala/org/apache/spark/sql/carbondata/execution/datasources/SparkCarbonFileFormat.scala]: ok

File: integration/spark-datasource/src/main/scala/org/apache/spark/sql/carbondata/execution/datasources/readsupport/SparkUnsafeRowReadSuport.scala (hunk 8 above)
Comment (ravipesala) [reply integration/spark-datasource/src/main/scala/org/apache/spark/sql/carbondata/execution/datasources/readsupport/SparkUnsafeRowReadSuport.scala]: ok

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/6354/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/8002/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6725/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/6366/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/8017/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6740/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/6370/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/8022/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6747/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/6376/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/6377/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/6378/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6757/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/8033/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/6384/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/8039/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6760/

**Pull request closed without merging**